samples/js/picasa/find-photos/src/main.js

```diff
@@ -0,0 +1,248 @@
+
+async function main(): Promise<void> {
+  const inSpec = PathSpec.parse(args._[0]);
+  const pinnedSpec = await inSpec.pin();
+  if (!pinnedSpec) {
+    throw `Invalid input dataset: ${inSpec.path.dataset}`;
+  }
+  const [db, input] = await pinnedSpec.value();
+  if (!input) {
+    throw `Invalid input spec: ${inSpec.toString()}`;
+  }
+  const outSpec = DatasetSpec.parse(args._[1]);
+  const [outDB, output] = outSpec.dataset();
+  const result = [];
+
+  await walk(input, db, async (v: any) => {
+    if (v instanceof Struct && isSubtype(sourceType, getTypeOfValue(v))) {
+      const w = parseInt(v['gphotoQ24width']['Q24t'], 10);
+      const h = parseInt(v['gphotoQ24height']['Q24t'], 10);
+
+      const photo: Object = {
+        sizes: await getSizes(v, w, h),
+        datePublished: getDate(Date.parse(v['published']['Q24t'])),
+        dateUpdated: getDate(Date.parse(v['updated']['Q24t'])),
+      };
+
+      if (isSubtype(hasTitle, v.type)) {
+        photo.title = v['title']['Q24t'];
+      }
+
+      if (isSubtype(hasExifTime, v.type)) {
+        if (v['gphotoQ24exiftime'].length === 1) {
+          photo.dateTaken = getDate(parseInt(
+            (await v['gphotoQ24exiftime'].get(0))['Q24t'], 10));
+        }
+      }
+
+      if (isSubtype(hasGeo, v.type)) {
+        const g = getGeo(v);
+        if (g) {
+          photo.geoposition = g;
+        }
+      }
+
+      if (isSubtype(hasFaces, v.type)) {
+        const f = await getFaces(v, w, h);
+        if (f) {
+          photo.faces = f;
+        }
+      }
+
+      result.push(newStruct('Photo', photo));
+      process.stdout.write(clearLine + `Indexed ${result.length} photos...`);
+      return true;
+    }
+  });
+
+  await outDB.commit(
+    output,
+    new Set(result),
+    {
+      meta: newStruct('', {
+        date: new Date().toISOString(),
+        input: pinnedSpec.toString(),
+      }),
+    }
+  );
+  await db.close();
+  await outDB.close();
+
+  process.stdout.write(clearLine);
+}
+
+function getGeo(input): ?Struct {
+  const pattern = /^(\-?\d+\.\d+) (\-?\d+\.\d+)$/;
+  const match = input['georssQ24where']['gmlQ24Point']['gmlQ24pos']['Q24t'].match(pattern);
+  if (!match) {
+    // This can happen even though we type-checked because the pattern might not match.
+    return null;
+  }
+
+  return newStruct('Geoposition', {
+    latitude: parseFloat(match[1]),
+    longitude: parseFloat(match[2]),
+  });
+}
+
+async function getSizes(input: Object, origWidth: number, origHeight: number)
+    : Promise<Map<Struct, string>> {
+  const thumbURL = (await input['mediaQ24group']['mediaQ24thumbnail']
+      .get(0))['url'].split('/');
+  const sizePart = thumbURL.length - 2;
+
+  const makeURL = s => {
+    thumbURL[sizePart] = `s${s}`;
+    return thumbURL.join('/');
+  };
+
+  const tuples = sizes.map(s => {
+    const r = fit(origWidth, origHeight, s, s);
+    if (r.scale > 1) {
+      // $FlowIssue: Does not understand that filter removes all null values.
```

Comment:
weird, but curious - what if you changed it from `filter(t => t)` to `filter(t => !!t)`? I remember having to do that in the past.